spynnaker/pyNN/models/neuron/structural_plasticity/synaptogenesis/formation/distance_dependent_formation.py

```diff
@@ -0,0 +1,163 @@
+# Copyright (c) 2017-2019 The University of Manchester
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import numpy
+from .abstract_formation import AbstractFormation
+from pacman.model.decorators.overrides import overrides
+from data_specification.enums import DataType
+
+
+class DistanceDependentFormation(AbstractFormation):
+    """ Formation rule that depends on the physical distance between neurons
+    """
+
+    __slots__ = [
+        "_grid",
+        "_p_form_forward",
+        "_sigma_form_forward",
+        "_p_form_lateral",
+        "_sigma_form_lateral",
+        "__ff_distance_probabilities",
+        "__lat_distance_probabilities"
+    ]
+
+    def __init__(
+            self, grid=numpy.array([16, 16]), p_form_forward=0.16,
+            sigma_form_forward=2.5, p_form_lateral=1.0,
+            sigma_form_lateral=1.0):
+        """
+
+        :param grid: (x, y) dimensions of the grid of distance
+        :param p_form_forward:\
+            The peak probability of formation on feed-forward connections
+        :param sigma_form_forward:\
+            The spread of probability with distance of formation on\
+            feed-forward connections
+        :param p_form_lateral:\
+            The peak probability of formation on lateral connections
+        :param sigma_form_lateral:\
+            The spread of probability with distance of formation on\
+            lateral connections
+        """
+        self._grid = numpy.asarray(grid, dtype=int)
+        self._p_form_forward = p_form_forward
+        self._sigma_form_forward = sigma_form_forward
+        self._p_form_lateral = p_form_lateral
+        self._sigma_form_lateral = sigma_form_lateral
```

Comment:
Should these be fully private i.e. __ rather than _?

Reply:
Done (and made some other variables private too)